Switch on/off all lights, and wait for the result

mjrider opened the pull request:
## Description:

This changes the order of the async tasks fired when switching lights on or of. This should make switching lights on a more parallel process

the current flow is( pseudo code)
```
foreach light in lights:
   fire change task
  wait change task
  fire state poll task
wait for all state poll tasks
```
the new flow will be:
```
foreach light in lights:
  fire change task
wait for allchange tasks

foreach light in lights:
  fire state poll task
wait for all state poll tasks
```

Pro's
when switching on a group of lights, there is no wait in firing the change task, which should make the time between lights switching on shorter

Con's
the state poll fire will wait for the last light in the group to finish the change task

**Related issue (if applicable):** fixes #17477.

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools: N/A
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

Only the order is changed, there should be no change in start/end scenarios, only the order in which the events are fired


[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

mjrider committed:
Switch on/off all lights, and wait for the result

Reuses the parallel_updates semaphore.
This is a small crutch which serializes platforms which already do tis
for updates. Platforms which can parallelize everything, this makes it
go faster

frenck commented:
@mjrider Please run `black --fast homeassistant` from the project root.
This will ensure the correct formatting of the code.

pvizeli reviewed (commented):
That look like the old code. We moved away because we hit device limits and eat our executors at once

mjrider commented:
> That look like the old code. We moved away because we hit device limits and eat our executors at once

any specifics on what those limits are, because it looks like this is bugging ppl to ( see issue )
We could fix it by chunking the operations, so a small group of lights trigger parallel, and larger groups a bit more serial

mjrider commented:
on a other note, i really do not understand what is wrong with the rflink test

pvizeli commented:
You can look into history of this file.

Some integration have a rate limit, i.e. hue or other serial device/protocol specific integration. We had initial all pararell like you change it now. But we run into so many issues that we going back to only use pararell for asyncio and seriel for sync device and later complete back to old handling before we was asyncio. There was to many issues with different device types.

balloob commented:
So the underlying issue here is indeed device capabilities. For everything but lights we use the service helper that will actually parallelize the requests per platform:

https://github.com/home-assistant/home-assistant/blob/9615ba3d99490e199a8f691a1be02626c04c5e25/homeassistant/helpers/service.py#L338-L343

For calling services we do it in sequence, but do the updates in parallel:

https://github.com/home-assistant/home-assistant/blob/9615ba3d99490e199a8f691a1be02626c04c5e25/homeassistant/helpers/service.py#L352

The reason we do the updates in parallel is because we actually have a semaphore that platforms can use to define the number of parallel requests it support (default for async is unlimited, sync is 1 at the same time). We don't have this currently for service calls / have one that is shared between updates/service calls. 

So to solve this problem correctly and once and for all, that's where we should look:

https://github.com/home-assistant/home-assistant/blob/9615ba3d99490e199a8f691a1be02626c04c5e25/homeassistant/helpers/entity_platform.py#L71

One solution could be that we also add `PARALLEL_REQUESTS`.

mjrider commented:
Oke, maybe this wasn't a good hacktoberfest issue to pickup ;-)

If i understand it correctly, the code in this PR highlights the a problem with the implementation of the platforms, which would make the system grind to a halt due to to many threads are waiting on serialized calls. For which it currently serializes the calls of turn_on/off in light?

Currently switching on a group of 100 lights would give you 100 lights switching on one by one
while doing 100 turn_on calls with each a different light would break home-assistant.

for the solution, if i understand balloob correctly:
- add a semaphore per platform, with `PARALLEL_REQUESTS` entries ( and default to `PARALLEL_UPDATES`  when not set ? ), so platforms can differentiate between updates and requests

- add a helper which is used in the `ToggleEntity`, to call turn_on/turn_off on a platform with a semaphore to throttle the amount of running task

- use that helper in locations like https://github.com/home-assistant/home-assistant/blob/82b77c2d29c95609d9ebdec0cbaf329cbb1d306d/homeassistant/helpers/entity.py#L569

pvizeli commented:
`PARALLEL_REQUESTS` with semaphore would be a good idea, ident. @balloob if do implement that correct, it should also solve some issue with threads/sync/multithreading.

mjrider commented:
@balloob @pvizeli i'm going to say challenge accepted, and trying to further pick up this issue. I hope you are willing to coach me at some point to make sure i'm understanding the architecture of home-assistant correctly.

i'll start with duplicating the parallel_update code to parallel_request in entity_platform.py and adding unittest for the new code

balloob commented:
Happy to help on this one. Anything to make things faster 👍 Hit me up on Discord for faster response time.

One other part that we miss now is that sometimes a platform is able to optimize requests if it knows work for multiple entities is coming in. Example, if I call `light.turn_on` with `entity_id: light.kitchen, light.living_room` and both these lights are the same platform, the platform might be able to send a command that targets two devices to the light hub/controller. But this might be out of scope for the current issue.

mjrider commented:
Improvement idea:
instead of using defining a new setting, which every integration should set, it make default to parallel_updates for the semaphore limit
PARALLEL_REQUEST = None ,  PARALLEL_UPDATES = None => both semaphore's are None
PARALLEL_REQUEST = None ,  PARALLEL_UPDATES = 2 => both semaphore's are 2
PARALLEL_REQUEST = 1 ,  PARALLEL_UPDATES = 2 => semaphore's are 1 , 2

that way a lot more platforms by default would give us the intended behaviour

balloob commented on homeassistant/components/light/__init__.py:
```diff
@@ -305,17 +306,22 @@ class SetIntentHandler(intent.IntentHandler):
-        if update_tasks:
-            await asyncio.wait(update_tasks)
+        if change_tasks:
+            await asyncio.wait(change_tasks)
```
We should use `gather`, so exceptions propagate correctly.

mjrider replied:
i keep it on gather, until the safe wrapper is available

pvizeli commented on homeassistant/helpers/entity.py:
```diff
@@ -551,6 +554,19 @@ def __repr__(self) -> str:
         return "<Entity {}: {}>".format(self.name, self.state)
 
+    # call an requests
+    async def async_request_call(self, request, params):
```
should be `async_request_call(self, coro, *args, **kwargs)`

mjrider replied:
I'm a bit stuck in the, and what are the changes to the rest of the code path, the caller, and the function call below. Could you help me with that
the whole `*` `**` magic in python is still a bit magic for me

mjrider replied:
I hope 1d170e6 is indeed what you asked

pvizeli replied:
Yep, perfect

mjrider committed:
Fix broken unittest

With manual validation, with help from @frenck, we found out that the
assertions are wrong and the test should be failing.

The sequence requested is
OFF
ON

without cancelation, this code should result in:
off,off,off,on,on,on

testable, by adding a `await hass.async_block_till_done()` between the
off and on call.

with cancelation. there should be less off call's so
off,on,on,on

mjrider commented:
I think the change is complete, so depending on the gather replacement, i think i can drop the draft state

balloob reviewed (approved):
I've had some discussion with @mjrider 1-on-1. I think that this is a perfect first step: reusing existing updates semaphore for control too.

I think that it is not a good idea to introduce a new semaphore, if the plan is to afterwards come up with a solution that can cover all different approaches that an integration can take. We should continue this conversation but probably in an architecture issue.

pvizeli reviewed (approved):
Some tests should be needed for `async_request_call` because it's a core code. Other is fine

balloob commented on homeassistant/helpers/entity.py:
```diff
@@ -551,6 +551,19 @@ def __repr__(self) -> str:
         return "<Entity {}: {}>".format(self.name, self.state)
 
+    # call an requests
+    async def async_request_call(self, coro, *args, **kwargs):
```
So I suggest that you actually pass in a coroutine. This currently is a coroutine function that produces a coroutine. Since coroutines are only executed once they are awaited, it's fine to have them pass in as a coroutine. 

```python
await light.async_request_call(light.async_turn_off(**off_pars))
```

pvizeli replied:
Yeah is better

balloob commented on homeassistant/helpers/entity.py:
```diff
@@ -551,6 +551,19 @@ def __repr__(self) -> str:
+            await self.parallel_updates.acquire()
+
+        try:
+            await coro(*args, **kwargs)
```
```suggestion
            await coro
```

pvizeli commented:
@balloob we could use a decorator for the entity object functions they do `async_request_call` functionality. So we reduce the complexity to use it over the codebase

pvizeli commented on homeassistant/components/light/__init__.py:
```diff
@@ -305,17 +306,22 @@ class SetIntentHandler(intent.IntentHandler):
-        if update_tasks:
-            await asyncio.wait(update_tasks)
+        if change_tasks:
+            await asyncio.gather(*change_tasks)
```
You should not use gather if you not can handle exception. That means he change the state but would never call async_update_ha_state if a integration fail. They messup the complete flow and end up in not stable core. That is very bad for core and HA. At this point I'm not ident with @balloob to mess up the stable execution

mjrider replied:
I changed it back to wait, i assumed that is what you wanted

mjrider committed:
Adding tests for async_request_call

mjrider committed:
Process review feedback

mjrider committed:
Switch gather with wait

mjrider committed:
:shirt: running black
